File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # how to calculate height difference?
        # we would use dfs
        # we dfs through each left and right branch
        # then we would calculate abs(left-right) <= 1
        balanced = True

        def dfs(node):
            nonlocal balanced

            if not node:
                return 0
            
            left_depth = dfs(node.left)
            right_depth = dfs(node.right)

            height_diff = abs(left_depth - right_depth)

            if height_diff > 1:
                balanced = False

            return 1 + max(left_depth, right_depth)
        
        dfs(root)
        return balanced

        

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int globalMax = INT_MIN;

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return globalMax;
    }

    int dfs(TreeNode* root) {
        if (root == nullptr) return 0;

        int leftValue = max(0, dfs(root->left));
        int rightValue = max(0, dfs(root->right));

        int valueOfPath = root->val + leftValue + rightValue;
        globalMax = max(globalMax, valueOfPath);

        return root->val + max(leftValue, rightValue);


    }
};
